email_account_validity/_base.py

```diff
@@ -142,19 +151,38 @@ async def generate_renewal_token(self, user_id: str) -> str:
             The generated string.
 
         Raises:
-            StoreError(500): Couldn't generate a unique string after 5 attempts.
+            SynapseError(500): Couldn't generate a unique string after 5 attempts.
         """
+        if not self._send_links:
+            # If the user isn't expected to click on a link, they're expected to enter a
+            # token manually into their client, which in turn sends the renewal request
+            # to the server, authenticated with an access token. This means that in this
+            # case we need the token to be shorter and less complex (hence the 8 digits
+            # string), but also that we don't need to make the token unique across the
+            # whole database.
+            renewal_token = random_digit_string(8)
+            await self._store.set_renewal_token_for_user(
+                user_id, renewal_token, unique=False,
+            )
+            return renewal_token
```

Comment:
it feels like it would be simpler to make this a separate function that can be called from `send_renewal_email_to_user` (which already has `if self._send_links` conditions)